account_invoice_start_end_dates/tests/test_invoice_start_end_dates.py

```diff
@@ -0,0 +1,94 @@
+        super().setUpClass()
+        cls.move_model = cls.env["account.move"]
+        cls.account_model = cls.env["account.account"]
+        cls.journal_model = cls.env["account.journal"]
+        cls.account_revenue = cls.env["account.chart.template"]._get_demo_account(
+            "income",
+            "income",
+            cls.env.company,
+        )
+        cls.sale_journal = cls.journal_model.search([("type", "=", "sale")], limit=1)
+        cls.maint_product = cls.env.ref(
+            "account_invoice_start_end_dates.product_maintenance_contract_demo"
+        )
+
+    def _date(self, date):
+        """convert MM-DD to current year date YYYY-MM-DD"""
+        return time.strftime("%Y-" + date)
+
+    def test_invoice(self):
+        invoice = self.move_model.create(
+            {
+                "date": self._date("01-01"),
+                "partner_id": self.env.ref("base.res_partner_2").id,
+                "journal_id": self.sale_journal.id,
+                "move_type": "out_invoice",
+                "invoice_line_ids": [
+                    (
+                        0,
+                        0,
+                        {
+                            "product_id": self.maint_product.id,
+                            "name": "Maintenance IPBX 12 mois",
+                            "price_unit": 2400,
+                            "quantity": 1,
+                            "account_id": self.account_revenue.id,
+                            "start_date": self._date("01-01"),
+                            "end_date": self._date("12-31"),
+                        },
+                    ),
+                    (
+                        0,
+                        0,
+                        {
+                            "product_id": self.maint_product.id,
+                            "name": "Maintenance téléphones 12 mois",
+                            "price_unit": 12,
+                            "quantity": 10,
+                            "account_id": self.account_revenue.id,
+                            "start_date": self._date("01-01"),
+                            "end_date": self._date("12-31"),
+                        },
+                    ),
+                    (
+                        0,
+                        0,
+                        {
+                            "product_id": self.maint_product.id,
+                            "name": "Maintenance Fax 6 mois",
+                            "price_unit": 120.75,
+                            "quantity": 1,
+                            "account_id": self.account_revenue.id,
+                            "start_date": self._date("01-01"),
+                            "end_date": self._date("06-30"),
+                        },
+                    ),
+                    (
+                        0,
+                        0,
+                        {
+                            "product_id": self.env.ref("product.product_product_5").id,
```

Comment:
looking at this test case i guess we test positive case here - all previous lines have dates set and we don't expect any error to be raised for product_product_5 because probably this product has no flag set - i'd maybe add some comments in this case + test negative case (where error should be raised)?

Reply:
`if moveline.start_date and not moveline.end_date:`
`if moveline.end_date and not moveline.start_date:`
```
if (
    moveline.end_date
    and moveline.start_date
    and moveline.start_date > moveline.end_date
):
```
hi @ajaniszewska-dev, I had the same question as you, I'm not sure what the author of this module intended, because the above conditions require at least 1 value to be assigned to `start_date` and `end_date`, so if no value is inputed, no error will be raised. I tried adding a filter in the lookup function like:
```
def _check_start_end_dates(self):
    for moveline in self.filtered("must_have_date"):
         ...
```
then the error will be raised, and I wonder if that change affects the original purpose of this module?

Reply:
@alexis-via  probably your input here would be valuable.

Reply:
I added the negative tests in my migration PR https://github.com/OCA/account-closing/pull/313